collect/src/main/kotlin/com/certora/collect/HashTreapMap.kt

```diff
@@ -280,6 +280,15 @@ internal class HashTreapMap<@Treapable K, V>(
         forEachPair { (k, v) -> h += AbstractMapEntry.hashCode(k, v) }
         return h
     }
+
+    override fun <R : Any> shallowMapReduce(map: (K, V) -> R, reduce: (R, R) -> R): R {
+        var result: R? = null
+        forEachPair {
+            val mapped = map(it.key, it.value)
+            result = result?.let { result -> reduce(result, mapped) } ?: mapped
+        }
+        return result!!
+    }
```

Comment:
is the type bound on `R` here just so we have nullability of `result`? I'm trying to think of a way to relax this, but I think you can only use, e.g., `lateinit` on non-null types.

Probably not worth it.

Reply:
We depend on this constraint in `mapReduceImpl`, where it's a bigger deal.  It avoids needing to allocate a result holder for each step of the traversal.